documentation/architecture/authorization/ABAC.md

```diff
@@ -0,0 +1,194 @@
+
+PDP and PIP can be abstracted and implemented as a separate system and even delegated
+to remote nodes.
+
+PEP is an integration point to convert existing request context into ABAC terms.
+
+For example in message flow authorization PEP is an `is_authorized` function.
+It takes a message and a worker state and creates an ABAC request which contains
+subject, action and resource attributes.
+
+### ABAC requests and policy rules
+
+ABAC request is a set of attributes derived from the current request and context by PEP.
+
+ABAC policy is a set of rules applied to the attributes.
+
+In order to make request-policy association easier, both contain ActionID, which
+describes a resource and an action being used.
+This allows for example to use it as a key in the database.
+
+For example ActionID for example above can be written as `Project/Update`
+
+#### Policy rules
+
+Policy rules for Ockam ABAC are defined as s-expressions.
+Each rule resolves into boolean value when "true" means "allow".
+
+##### Bool rule:
+
+- `true`
+- `false`
+
+Mostly useful in logical rules, but can be used as allow-all or deny-all policies
+
+##### Comparison rules:
+
+- equality `(= subject.foo "bar")` `(!= subject.foo "bar")`
+- numeric comparison `(> subject.foo 1)`, `(< subject.foo 3)`
+
+Supports operators `=` `!=` `>` and `<`
+First argument MUST be an Attribute
+Second argument can be an Attribute or a Value
+
+##### Inclustion rule:
+
+Matches that an attribute belongs to a list.
+
+- `(member? subject.key [1 2 3])`
+- `(member? subject.key resource.keys)`
+- `(member? "foo" action.keys)`
+
+First argument can be an Attribute or a Value
+Second argument can be an Attribute or a List
+If second argument is an Attribute, the corresponding Attribute should resolve to a List
+
+##### Logical rules:
+
+Combine other rules
+
+- `(not (in "foo" action.keys))`
+- `(and (= subject.key, "foo") (> subject.things 2))`
+- `(or (= subject.key, "foo") (> subject.things 2) (= action.thing "smth"))`
+- `(if (= subject.role "member") (in subject.name resource.members) (= action.special true))`
+
+`not` - unary operator
+`if` - ternary operator with condition rule, true rule and false rule
+  if condition rule matches - true rule is checkes
+  if condition rule doesn't match - false rule is checked
+`and|or` - operators combining multiple rules (can be 2 and more)
+
+##### Attributes
+
+- `subject.key`
+- `action.other_key`
+- `resource.something`
+
+Attribute consists of type and name.
+Type can be subject, action or resource
+Name can contain lowercase ASCII letters, numbers and `_`
+
+##### Values
+
+- `"foo"`
+- `1`
+- `0.35`
+- `true`
+
+Values can be strings, numbers or booleans.
+
+##### Lists
+
+- `[1 2 3]`
+- `["foo" "bar"]`
+
+List consists of multiple values.
+
+**Attributes, values and lists can be arguments in rules, but cannot be rules,
+because they don't resolve to boolean**
```

Comment:
Except for the boolean values.

Reply:
`true` and `false` are rules, attributes with boolean values should not be rules because we can't guarantee attribute types.